Composable search with form

nicolethoen opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #6212 

I pulled out the dynamically built advanced search into a separate (unexposed) component.
I updated the docs for the dynamically built advanced search example for clarity.
I added a demo to demonstrate building a custom composable advanced search form.

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-6279.surge.sh

CooperRedhat commented:
This is some super complex functionality and everything seems to be working well! Awesome. 

I clicked around for a while and here is what I noticed:
- [Original issue](https://github.com/patternfly/patternfly-react/issues/6212) shows saving search state using chips, examples are using strings. Is that going to be added later, or was it dropped?
- In the demo: 
  - the advanced search keeps the values of form elements when opening/closing the `Popper`, with exception of the actual date value
  - After filling in values for `has the words`, `Date`, and `Date within`, the string that represents the search state in the input box above only shows the `has the words` value, but the example shows all of the delimited search parameters.
  - There seems to be a lot of work handed to the user in terms of handling focus, stopping event propagation, etc. So it's great to have this complex demo for them. In the long term is there a way to simplify that interface? Or could you provide some documentation in the demo code to explain why/how you handle focus, key events, etc?

CooperRedhat reviewed (commented):
See comment above

kmcfaul reviewed (commented):
The submit button causes a null warning in the console for the click handler, but looks great for everything else! Awesome work making it composable.

nicolethoen commented:
@CooperRedhat we will need to build a new component to allow for the 'input with chips look' and that is on the road map. So this story is just to demonstrate that advanced search forms can be built composably. I did try to add more comments within the demo code, as well as a few notes above the demo. Do you think that's helpful? I think I've addressed the rest of your comments

I added the after:date and to:date to the input if it's supplied in the advanced form, and then if you begin manually typing into the input still only updates the 'has words' field since I don't think it's worth the time/effort to be building a parser for the words 'after:' and 'to:' and more date parsing to this simple/small demo(?) @mcarrano Let me know what you think.

CooperRedhat reviewed (dismissed):
LGTM

CooperRedhat commented:
> @CooperRedhat we will need to build a new component to allow for the 'input with chips look' and that is on the road map. So this story is just to demonstrate that advanced search forms can be built composably. I did try to add more comments within the demo code, as well as a few notes above the demo. Do you think that's helpful? I think I've addressed the rest of your comments
> 
> I added the after:date and to:date to the input if it's supplied in the advanced form, and then if you begin manually typing into the input still only updates the 'has words' field since I don't think it's worth the time/effort to be building a parser for the words 'after:' and 'to:' and more date parsing to this simple/small demo(?) @mcarrano Let me know what you think.

Yes the comments definitely help give some context to what's going on. Should help our users modify your example more easily to fit their needs! nice

mcarrano reviewed (commented):
@nicolethoen this is definitely better. I still had a bit of a problem figuring out the logic of how the 'Date within' and 'Date' fields were being used. Could we make a slight change to the labeling on the form to be something like "Date within" for the first field and "of" in front of the Date picker? So the whole thing reads like "date within 2 weeks of 09/09/2021" for example. I think that would clarify the example a bit. Thanks.

mcarrano reviewed (approved):
Looks great. Thanks @nicolethoen !

CooperRedhat reviewed (approved):
LGTM

tlabaj reviewed (changes requested):
Great work!  this looks awesome.  Just a few comments.

tlabaj commented on packages/react-core/src/components/SearchInput/AdvancedSearchMenu.tsx:
```diff
@@ -0,0 +1,223 @@
+  /** Function called to toggle the advanced search menu */
+  onToggleAdvancedMenu?: (e: React.SyntheticEvent<HTMLButtonElement>) => void;
+  /** Flag for toggling the open/close state of the advanced search menu */
+  showSearchMenu?: boolean;
```
Maybe  `isSearchMenuOpen`

kmcfaul commented on packages/react-core/src/components/SearchInput/examples/SearchInput.md:
```diff
@@ -194,6 +194,12 @@ TextInputSelectAll = () => {
 
 ### Advanced
 
+The search input component can be used to dynamically build a one to one attribute-value advanced search.
```
Same!

kmcfaul reviewed (approved):
I agree with Titani that `isSearchMenuOpen` might be a bit more consistent. Everything else looks great! Love the descriptions/comments in the docs.

tlabaj reviewed (approved):
looks great!

mturley reviewed (approved):
This looks awesome!

mcoker commented on packages/react-core/src/demos/SearchInput.md:
```diff
@@ -0,0 +1,299 @@
+  const advancedForm = (
+    <div ref={advancedSearchPaneRef} role="dialog" aria-label="Advanced search form">
+      <Form>
+        <Card>
```
The card component doesn't match the search input styling. Can we expose the `.pf-c-search-input__menu` and `.pf-c-search-input__menu-body` elements and use those instead? https://github.com/patternfly/patternfly-react/blob/06b7d17f1b48996f17fd1021f1834eaeebd7bc18/packages/react-core/src/components/SearchInput/SearchInput.tsx#L358

nicolethoen replied:
In this case, the menu is not a child of search input - so the styles are not applied correctly to the form if they are in orphaned `.pf-c-search-input__menu` and `.pf-c-search-input__menu-body`components.

mcoker replied:
Hmm... I'm not sure I understand. The rendered markup all seems to support the styling working, I just renamed the card elements to search input menu/menu-body and moved the actions into the form as an action group:

<img width="674" alt="Screen Shot 2021-09-10 at 1 38 04 PM" src="https://user-images.githubusercontent.com/35148959/132901707-47c546d3-fa16-40b8-af56-52a224504a1e.png">

mcoker commented on packages/react-core/src/demos/SearchInput.md:
```diff
@@ -0,0 +1,299 @@
+              </FormGroup>
+            </Grid>
+          </CardBody>
+          <CardFooter>
```
The spacing for card footer actions doesn't match the spacing for form actions. If we swap the card out with the search input menu/menu-body elements, we can use a form action group for the right spacing here. If we aren't going to do that, I think this form should just be in a `<CardBody>`, and the form action group used for the actions.

nicolethoen replied:
I can do that!

nicolethoen replied:
well - actually.  I will play with it - I think i was getting into sticky territory where the actions need to be in a `Form` and so the `Form` was wrapping the `CardBody` and `CardFooter`. I'll take another stab at it

mcoker reviewed (changes requested):
Looks good, left a couple of comments.

tlabaj reviewed (dismissed):
Still looks good!

jessiehuff reviewed (commented):
Looks pretty good to me! The only weird thing with the demo is that when I'm using keyboard (or screen reader), and I select something for the "Date within" option, the focus is completely lost, and I have to navigate all the way back to it which would be pretty frustrating for keyboard and screen reader users. I'm guessing this is just a demo issue though. :)

nicolethoen commented:
good catch, @jessiehuff  I'll fix that!

mcarrano reviewed (approved):
Looks good. Thanks @nicolethoen !

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@4.12.66
 - @patternfly/react-code-editor@4.3.53
 - @patternfly/react-console@4.11.34
 - @patternfly/react-core@4.156.0
 - @patternfly/react-docs@5.21.32
 - @patternfly/react-inline-edit-extension@4.7.76
 - demo-app-ts@4.120.3
 - @patternfly/react-log-viewer@4.6.6
 - @patternfly/react-table@4.29.69
 - @patternfly/react-topology@4.9.72
 - @patternfly/react-virtualized-extension@4.9.41

Thanks for your contribution! :tada:
